library/autodiff/examples/sin.rs

```diff
@@ -1,22 +1,50 @@
 use autodiff::autodiff;
 
+// This syntax defines a new function called `cos_inplace` by differentiating the code below.
 #[autodiff(cos_inplace, Reverse, Const)]
 fn sin_inplace(#[dup] x: &f32, #[dup] y: &mut f32) {
     *y = x.sin();
 }
 
+// This syntax (with only a function declaration) differentiates `sin_inplace`
+// and makes the result callable as `cos_inplace_fwd`.
+//
+// Const refers to the (non-existent in this case) return value. In forward
+// mode, we duplicate the inputs x and read the result from dy, without a need
+// for y (which is undefined because we use `dup_noneed`).
+#[autodiff(sin_inplace, Forward, Const)]
+fn cos_inplace_fwd1(#[dup] x: &f32, dx: &f32, #[dup_noneed] y: &mut f32, dy: &mut f32);
+
+// This has identical semantics to above, but specifies dup semantics up-front
+// instead of using inline attribute macros.
+#[autodiff(sin_inplace, Forward, Const, Duplicated, DuplicatedNoNeed)]
+fn cos_inplace_fwd2(x: &f32, dx: &f32, y: &mut f32, dy: &mut f32);
 
 fn main() {
     // Here we can use ==, even though we work on f32.
     // Enzyme will recognize the sin function and replace it with llvm's cos function (see below).
     // Calling f32::cos directly will also result in calling llvm's cos function.
-    let a = 3.1415;
+    let a = std::f32::consts::PI / 6.0;
     let mut da = 0.0;
     let mut y = 0.0;
     cos_inplace(&a, &mut da, &mut y, &mut 1.0);
 
     dbg!(&a, &da, &y);
     assert!(da - f32::cos(a) == 0.0);
+
+    // To use forward mode, we specify da and recover the result from dy. The
+    // result value of y is undefined due to DuplicatedNoNeed.
+    let mut dy = 0.0;
+    da = 1.0;
+    y = 0.0;
+    cos_inplace_fwd1(&a, &da, &mut y, &mut dy);
+    dbg!(&a, &da, &dy);
+    assert!(dy - f32::cos(a) == 0.0);
+
+    // Same as above using the second syntax
+    cos_inplace_fwd2(&a, &da, &mut y, &mut dy);
+    dbg!(&a, &da, &dy);
+    assert!(dy - f32::cos(a) == 0.0);
```

Comment:
@ZuseZ4  I think we have an issue here (```called `Option::unwrap()` on a `None` value```). This test passes in `debug` (`cargo +enzyme test --example sin`), but fails with `--release` (opposite the usual pattern). Moreover, if *either* these three lines are removed or line 40-42 above, it also passes with `--release`.
[rustc-ice-2023-11-06T02_55_49-174175.txt](https://github.com/EnzymeAD/rust/files/13261991/rustc-ice-2023-11-06T02_55_49-174175.txt)

Reply:
I wired up some rust error handler into the autodiff code, now you should get a proper rust failure:
```
error: failed to prepare AutoDiff: src: _ZN4sin223primal_cos_inplace_fwd217h4e079b7a356a5d44E, target: _ZN4sin216cos_inplace_fwd217he70028a71815c282E, could not find src function
```

Can you please run the example with `ENZYME_PRINT_MOD=1` (see https://github.com/EnzymeAD/rust#enzyme-config) and create a new issue? I think the reason might be that the function is turned into a global. There might be a function to "step" through globals if you have some time to debug, otherwise I can try to check once I am done with moving away from the macro to a compiler pass.

Reply:
I don't know how to start making sense of the 21 MB of output, but created an issue: https://github.com/EnzymeAD/Enzyme/issues/1555